### ISSUE TITLE
Gstreamer (and examples) set up on Ubuntu 16.04 <case:445081>

### ISSUE THREAD

**Issue by jveitchmichaelis** [TheImagingSource/tiscamera#150]
I've followed the steps in the readme - all dependencies are installed. I've run `sudo make install` and which has put things in `/usr/`. `tcam-ctrl` works, I've set up the uvc extensions, and `qv4l2` works. I've not been able to get any of the examples working on this machine running Ubuntu 16.04 LTS.

Seems to boil down to not setting gstreamer up properly, but I'm not familiar enough with it to really know what I'm looking for. I've tried this from the wiki:

`GST_PLUGIN_PATH=:/home/<user>/tis/tiscamera/src:/home/<user>/.$:/home/<user>/tis`

However the examples on that page reference gstreamer-0.1 and this export (with the right path) doesn't seem to do anything (checked with `gst-inspect-1.0 | grep tcambin`). 

Stuff is being installed to `/usr/lib/x86_64-linux-gnu/gstreamer-1.0` which seems to have a lot of other Gstreamer modules which are loaded fine, but I think something is causing the plugin loader to fail. In a new terminal:

```
ls /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcam*
/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamautoexposure.so
/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamautofocus.so
/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcambin.so
/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamsrc.so
/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamwhitebalance.so
```

```
gst-launch-1.0 tcambin ! videoconvert ! ximagesink

(gst-plugin-scanner:9953): GStreamer-WARNING **: Failed to load plugin '/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamwhitebalance.so': /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamwhitebalance.so: undefined symbol: _ZN4tcam8Property9set_valueERKlb

(gst-plugin-scanner:9953): GStreamer-WARNING **: Failed to load plugin '/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamautofocus.so': /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamautofocus.so: undefined symbol: _ZN4tcam8Property9set_valueERKlb
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc

(gst-plugin-scanner:9958): GStreamer-WARNING **: Failed to load plugin '/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamautofocus.so': /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamautofocus.so: undefined symbol: _ZN4tcam8Property9set_valueERKlb

(gst-plugin-scanner:9963): GStreamer-WARNING **: Failed to load plugin '/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamsrc.so': /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamsrc.so: undefined symbol: _ZNK4tcam12MemoryBuffer15get_buffer_sizeEv

(gst-plugin-scanner:9968): GStreamer-WARNING **: Failed to load plugin '/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamsrc.so': /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcamsrc.so: undefined symbol: _ZNK4tcam12MemoryBuffer15get_buffer_sizeEv
WARNING: erroneous pipeline: no element "tcambin"
```
```
/usr/lib/x86_64-linux-gnu/gstreamer-1.0$ gst-inspect-1.0 ./libgsttcambin.so
Plugin Details:
  Name                     tcambin
  Description              Tcam Video Bin
  Filename                 /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgsttcambin.so
  Version                  1.0.0
  License                  Proprietary
  Source module            tcam
  Binary package           tcambin
  Origin URL               theimagingsource.com

  tcambin: Tcam Video Bin

  1 features:
  +-- 1 elements
```

If I do:

```
export GI_TYPELIB_PATH="/home/tiscamera/build/src/gobject/:${GI_TYPELIB_PATH}"
export GST_PLUGIN_SYSTEM_PATH_1_0="/home/tiscamera/build/src/gstreamer-1.0/:${GST_PLUGIN_SYSTEM_PATH_1_0}"
```
Then Gstreamer doesn't complain about the module, but the basic example doesn't work:

```
$ gst-launch-1.0 tcambin ! videoconvert ! ximagesink
Setting pipeline to PAUSED ...
ERROR: Pipeline doesn't want to pause.
Setting pipeline to NULL ...
Freeing pipeline ...
```

Also with `live_stream.sh` (I've adjusted the frame size to be correct for the camera):

```
~/tiscamera/examples/gstreamer-1.0$ ./live-stream.sh 39410177
Setting pipeline to PAUSED ...
ERROR: Pipeline doesn't want to pause.
Setting pipeline to NULL ...
Freeing pipeline ...
```

`tcam_view` loads, but shows only blank windows and empty dropdown menus. Similarly with the C files, none of the examples show anything when I run them.

Even this is one is a bit weird, but as everything seems to be built on gstreamer it's failing too and reporting a generic "no device" rather than the pipeline itself failing.

```
~/tiscamera/examples/cpp/build/examples$ tcam-ctrl -l; ./tcam_demo
Available devices:
Model		Type	Serial

DFK 33UP2000	V4L2	25710592

No cameras found.
```

This pipeline delivers an image fine:
```
gst-launch-1.0 v4l2src device="/dev/video0" ! videoconvert ! ximagesink
Setting pipeline to PAUSED ...
Pipeline is live and does not need PREROLL ...
Setting pipeline to PLAYING ...
New clock: GstSystemClock
ERROR: from element /GstPipeline:pipeline0/GstXImageSink:ximagesink0: Output window was closed
Additional debug info:
ximagesink.c(700): gst_x_image_sink_handle_xevents (): /GstPipeline:pipeline0/GstXImageSink:ximagesink0
Execution ended after 0:00:03.211581206
Setting pipeline to PAUSED ...
Setting pipeline to READY ...
Setting pipeline to NULL ...
Freeing pipeline ...
```

**Comment by Jangelaw** [comment]
I met the same problem!!! Did you solve it or not?

**Comment by wazawoo** [comment]
Yes same here

**Comment by jveitchmichaelis** [comment]
Hi Stefan, 

The problem seems to be down to gstreamer not finding the `tcambin` element, I get the same error:

` ERROR: pipeline could not be constructed: no element "tcambin" `

I'm going to try installing in a docker container, perhaps my gstreamer install is messed up somewhere.

Thanks

**Comment by TIS-Stefan** [comment]
Hello

If the "tcambin" is not found, then you either did not compile and install the modules as shown in the readme.md at https://github.com/TheImagingSource/tiscamera or you may changed the GStreamer search path, which is not necessary.

Stefan
--

**Comment by TIS-Stefan** [comment]
This is a test email to tiscamera issues.

Stefan

**Comment by jveitchmichaelis** [comment]
Managed to get this working, with some offline help from Stefan.

My steps:

1) Run `sudo cmake uninstall` and remove the build directory
2) Run `apt-get update` and `upgrade`, do a reinstall of gstreamer
3) Remove the gstreamer cache (in `~/.cache/gstreamer-1.0` - I nerfed the whole folder)
4) `git pull`
5) Added lines to CMakeLists.txt to prevent CMAKE stripping the RPATH of the libraries (and to set it to /usr/lib). https://stackoverflow.com/a/32470070/395457 
5) Make a build folder and install as normal,
6) Opened a new terminal window to check it works

Seems like most of the problem was libraries not being correctly located, could possibly also have helped to pull some recent updates.
